subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareResourceHasher.java

```diff
@@ -0,0 +1,199 @@
+        delegate.appendConfigurationToHasher(hasher);
+        hasher.putString(getClass().getName());
+        hasher.putString(lineEndingSensitivity.name());
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(RegularFileSnapshotContext snapshotContext) throws IOException {
+        return lineEndingSensitivity.isCandidate(snapshotContext.getSnapshot()) ?
+            hashContent(snapshotContext) :
+            delegate.hash(snapshotContext);
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(ZipEntryContext zipEntryContext) throws IOException {
+        return lineEndingSensitivity.isCandidate(zipEntryContext) ?
+            hashContent(zipEntryContext) :
+            delegate.hash(zipEntryContext);
+    }
+
+    @Nullable
+    private HashCode hashContent(RegularFileSnapshotContext snapshotContext) throws IOException {
+        try {
+            return hashContent(new File(snapshotContext.getSnapshot().getAbsolutePath()));
+        } catch (BinaryContentDetectedException e) {
+            return delegate.hash(snapshotContext);
+        }
+    }
+
+    @Nullable
+    private HashCode hashContent(ZipEntryContext zipEntryContext) throws IOException {
+        return zipEntryContext.getEntry().withInputStream(inputStream -> {
+            try {
+                return hashContent(inputStream);
+            } catch (BinaryContentDetectedException e) {
+                return delegate.hash(zipEntryContext);
+            }
+        });
+    }
+
+    private HashCode hashContent(InputStream inputStream) throws BinaryContentDetectedException, IOException {
+        return new LineEndingAwareInputStreamHasher().hash(inputStream);
+    }
+
+    private HashCode hashContent(File file) throws BinaryContentDetectedException, IOException {
+        try (FileInputStream inputStream = new FileInputStream(file)) {
+            return hashContent(inputStream);
+        }
+    }
+
+    private static class BinaryContentDetectedException extends Exception {
+    }
+
+    @VisibleForTesting
+    static class LineEndingAwareInputStreamHasher {
+        private static final HashCode SIGNATURE = Hashing.signature(LineEndingAwareInputStreamHasher.class);
+        int peekAhead = -1;
+
+        public HashCode hash(InputStream inputStream) throws IOException, BinaryContentDetectedException {
+            byte[] buffer = new byte[8192];
+            PrimitiveHasher hasher = Hashing.newPrimitiveHasher();
+
+            hasher.putHash(SIGNATURE);
+            while (true) {
+                int nread = read(inputStream, buffer);
+                if (nread < 0) {
+                    break;
+                }
+
+                if (checkForControlCharacters(buffer, nread)) {
+                    throw new BinaryContentDetectedException();
+                }
+
+                hasher.putBytes(buffer, 0, nread);
+            }
+            return hasher.hash();
+        }
+
+        int read(InputStream inputStream, byte[] b) throws IOException {
```

Comment:
Should this be `private` or `@VisibleForTesting`?

Reply:
`@VisibleForTesting` - added.